django/http/request.py

```diff
@@ -97,39 +97,48 @@ def _set_content_type_params(self, meta):
             else:
                 self.encoding = self.content_params['charset']
 
-    def _get_raw_host(self):
-        """
-        Return the HTTP host using the environment or request headers. Skip
-        allowed hosts protection, so may return an insecure host.
-        """
-        # We try three options, in order of decreasing preference.
-        if settings.USE_X_FORWARDED_HOST and (
-                'HTTP_X_FORWARDED_HOST' in self.META):
-            host = self.META['HTTP_X_FORWARDED_HOST']
+    @cached_property
+    def _parsed_host_header(self):
+        use_x_fw_host = settings.USE_X_FORWARDED_HOST
+        use_x_fw_port = settings.USE_X_FORWARDED_PORT
+
+        port_in_x_fw_host = False
+        default_port = ('443' if self.is_secure() else '80')
+
+        if use_x_fw_host and 'HTTP_X_FORWARDED_HOST' in self.META:
+            host, port = _parse_host_header(self.META['HTTP_X_FORWARDED_HOST'])
+            port_in_x_fw_host = port != ''
         elif 'HTTP_HOST' in self.META:
-            host = self.META['HTTP_HOST']
+            host, port = _parse_host_header(self.META['HTTP_HOST'])
         else:
             # Reconstruct the host using the algorithm from PEP 333.
-            host = self.META['SERVER_NAME']
-            server_port = self.get_port()
-            if server_port != ('443' if self.is_secure() else '80'):
-                host = '%s:%s' % (host, server_port)
-        return host
+            host, port = self.META['SERVER_NAME'], str(self.META['SERVER_PORT'])
+            if port == default_port:
+                port = ''
+
+        if use_x_fw_port and 'HTTP_X_FORWARDED_PORT' in self.META:
+            if port_in_x_fw_host:
+                raise ImproperlyConfigured('HTTP_X_FORWARDED_HOST contains a port number '
+                                           'and USE_X_FORWARDED_PORT is set to True')
+            port = self.META['HTTP_X_FORWARDED_PORT']
+
+        reconstructed = '%s:%s' % (host, port) if port else host
```

Comment:
to match the behavior of `_get_raw_host()` I think this reconstruction should only occur in the `SERVER_NAME` scenario.

Reply:
I don't think so, `_get_raw_host` returned the `Host`-header as is: https://github.com/django/django/blob/master/django/http/request.py#L110 

So if `_get_raw_host` passed in a port it would have been returned again. This is one of the reasons why the patch looks rather tricky.